File: Implementations/09 - Range Queries (2)/09.2 - 1D Range Queries (3)/BIT.cpp
/**
* Description: 1D range sum query with point update
* Source: TopCoder
* Verification: SPOJ Fenwick
*/

template<class T, int SZ> struct BIT {
    T bit[SZ+1];
    
    BIT() { memset(bit,0,sizeof bit); }
    
    void upd(int k, T val) { // add val to index k
        for( ;k <= SZ; k += (k&-k)) bit[k] += val;
    }
    
    T query(int k) {
        T temp = 0;
        for (;k > 0;k -= (k&-k)) temp += bit[k];
        return temp;
    }
    T query(int l, int r) { return query(r)-query(l-1); } // range query [l,r]
};

File: Implementations/content/various/BIT.kt
class BIT() {
    val v = LongArray(SZ)
    fun upd(_a: Int, b: Long) {
        var a = _a
        while (a < SZ) {
            v[a] += b
            a += (a and -a)
        }
    }
    fun query(_a: Int): Long {
        var a = _a
        var ans: Long = 0
        while (a > 0) {
            ans += v[a]
            a -= (a and -a)
        }
        return ans
    }
}